src/charm.py

```diff
@@ -97,6 +104,12 @@ def __init__(self, *args):
         self._database = DatabaseRequires(
             self, relation_name="database", database_name=DATABASE_NAME
         )
+        # Setting attributes to detect broken relations until
+        # https://github.com/canonical/operator/issues/940 is fixed
+        self._database_relation_breaking = False
```

Comment:
Hi @benhoyt,

According to Juju documentation https://juju.is/docs/sdk/constructs#heading--framework,
The attributes such as `self._database_relation_breaking` should be set to their default values upon every hook execution. However, I tested it  and those states are kept somehow. 
I both modified the integration test and tested manually by adding additional config-changed hook to see the effect. 
After relation broken, running a config-changed hook did not reset the init attribute state which conflicts with the documentation. If the documentation was correct, the integration tests should fail in this PR but they are not failing (?).
Please enlighten us.

Reply:
@gatici First, there's a discrepancy here between Harness tests (with the same Harness instance) and reality. The charm instance isn't reinitialized between hook executions when the Harness triggers events.

This also applies to deferred, which may be what you're seeing here? Normally the charm is a completely new instance (with attributes reset by `__init__`) every hook execution, but if an event was deferred, the deferred events are run first against the charm instance, and then the "main" event also executed on that same charm instance (without resetting attributes).

These discrepancies are a problem that we hope to address that this cycle. See https://github.com/canonical/operator/issues/736

So I wonder if the difference @dariofaccin is seeing is because there were no deferred events in that case (for whatever reason).

CC: @tonyandrewmeyer for visibility

Reply:
@benhoyt  Many thanks for your help and quick response!